swift/Arrow/Sources/Arrow/ArrowCExporter.swift

```diff
@@ -0,0 +1,110 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+import ArrowC
+import Atomics
+
+extension String {
+    var cstring: UnsafePointer<CChar> {
+        (self as NSString).cString(using: String.Encoding.utf8.rawValue)!
+    }
+}
+
+// The memory used by UnsafeAtomic is not automatically
+// reclaimed. Since this value is initialized once
+// and used until the program/app is closed it's
+// memory will be released on program/app exit
+let exportDataCounter: UnsafeAtomic<Int> = .create(0)
+
+public class ArrowCExporter {
+    private class ExportData {
+        let id: Int
+        var cArray = ArrowC.ArrowArray()
+        private let arrowData: ArrowData
+        private(set) var data = [UnsafeRawPointer?]()
+        private(set) var buffers: UnsafeMutablePointer<UnsafeRawPointer?>
+        init(_ arrowData: ArrowData) {
+            id = exportDataCounter.loadThenWrappingIncrement(ordering: .relaxed)
+
+            // keep a reference to the ArrowData
+            // obj so the memory doesn't get
+            // deallocated
+            self.arrowData = arrowData
+            for arrowBuffer in arrowData.buffers {
+                data.append(arrowBuffer.rawPointer)
+            }
+
+            self.buffers = UnsafeMutablePointer(mutating: data)
+            ArrowCExporter.exportedData[id] = self
+        }
+
+        func release() {
+            // the data associated with this export data
+            // does not need to be released as they are
+            // still associated with the ArrowBuffer
+            // and it will deallocate this memory
```

Comment:
The `release` pointer **is** responsible for releasing/deallocating any data backing the array.
If you're doing something different, then you're not respecting the C Data Interface's contract.

Which means you probably want to keep the `ArrowBuffer`s stored in `ExportData`.

Reply:
We keep a reference to the arrowdata object which contains the buffers so the buffers are currently being stored until released.

Reply:
Thanks. I was confused by the comment, can you perhaps make it less misleading?

Reply:
Will do.